File: modules/boost/simd/bitwise/script/lo.py
[ ## this file was manually modified by jt
    {
     'functor' : {
         'description' : ['The function returns the low part of the entry',
                          'as an integer of the same size as the input.'],
         'return' : ['always returns an integer value'],
         'module' : 'boost',
         'arity' : '1',
         'call_types' : [],
         'ret_arity' : '0',
         'rturn' : {
             'default' : 'dtype',
            },
         'type_defs' : ['typedef typename boost::dispatch::meta::as_integer<T,unsigned>::type ir_t;',
                        'typedef typename boost::dispatch::meta::downgrade<ir_t>::type dtype;',
                        'typedef typename boost::simd::meta::scalar_of<ir_t>::type scal;'],
         'types' : ['real_', 'int64_', 'int32_', 'int16_'],
         'simd_types' : ['real_', 'sintgt_8_'],
        },
     'info' : 'manually modified',
     'unit' : {
         'global_header' : {
             'first_stamp' : 'created  by jt the 18/02/2011',
             'included' : ['#include <boost/dispatch/meta/downgrade.hpp>'],
             'no_ulp' : 'True',
             'notes' : [],
             'stamp' : 'modified by jt the 18/02/2011',
            },
         'ranges' : {
             'default' : [['T(0)', 'T(10)']],
             'sintgt_8_' :  [['T(0)', 'T(10)']],
            },
         'specific_values' : {
             'default' : {
                 'boost::simd::One<T>()' : {'result' : 'boost::simd::One<dtype>()','ulp_thresh' : '0.5',},
                 'boost::simd::Zero<T>()' : {'result' : 'boost::simd::Zero<dtype>()','ulp_thresh' : '0.5',},
                },
             'real_' : {
                 'boost::simd::Nan<T>()' : {'result' : 'boost::simd::Mone<dtype>()','ulp_thresh' : '0.5',},
                 'boost::simd::One<T>()' : {'result' : 'boost::simd::Zero<dtype>()','ulp_thresh' : '0.5',},
                 'boost::simd::Zero<T>()' : {'result' : 'boost::simd::Zero<dtype>()','ulp_thresh' : '0.5',},
                },
            },
         'verif_test' : {
             'property_call' : {
                },
             'property_value' : {
                },
             'simd' : {
                },
             'ulp_thresh' : {
                },
            },
        },
    },
]

File: modules/boost/simd/bitwise/unit/scalar/lo.cpp
//==============================================================================
//         Copyright 2003 - 2012   LASMEA UMR 6602 CNRS/Univ. Clermont II
//         Copyright 2009 - 2012   LRI    UMR 8623 CNRS/Univ Paris Sud XI
//
//          Distributed under the Boost Software License, Version 1.0.
//                 See accompanying file LICENSE.txt or copy at
//                     http://www.boost.org/LICENSE_1_0.txt
//==============================================================================
#define NT2_UNIT_MODULE "nt2 boost.simd.bitwise toolbox - lo/scalar Mode"

//////////////////////////////////////////////////////////////////////////////
// unit test behavior of boost.simd.bitwise components in scalar mode
//////////////////////////////////////////////////////////////////////////////
/// created  by jt the 18/02/2011
///
#include <boost/simd/toolbox/bitwise/include/functions/lo.hpp>
#include <boost/simd/sdk/simd/native.hpp>
#include <boost/dispatch/meta/downgrade.hpp>

#include <boost/type_traits/is_same.hpp>
#include <boost/dispatch/functor/meta/call.hpp>
#include <nt2/sdk/unit/tests.hpp>
#include <nt2/sdk/unit/module.hpp>
#include <boost/simd/toolbox/constant/constant.hpp>


NT2_TEST_CASE_TPL ( lo_real__1_0,  BOOST_SIMD_REAL_TYPES)
{

  using boost::simd::lo;
  using boost::simd::tag::lo_;
  typedef typename boost::dispatch::meta::as_integer<T,unsigned>::type ir_t;
  typedef typename boost::dispatch::meta::downgrade<ir_t>::type dtype;
  typedef typename boost::simd::meta::scalar_of<ir_t>::type scal;
  typedef typename boost::dispatch::meta::as_integer<T>::type iT;
  typedef typename boost::dispatch::meta::call<lo_(T)>::type r_t;
  typedef typename boost::simd::meta::scalar_of<r_t>::type sr_t;
  typedef typename boost::simd::meta::scalar_of<r_t>::type ssr_t;
  typedef dtype wished_r_t;


  // return type conformity test
  NT2_TEST( (boost::is_same < r_t, wished_r_t >::value) );
  std::cout << std::endl;

  // specific values tests
  NT2_TEST_EQUAL(lo(boost::simd::Nan<T>()), boost::simd::Mone<dtype>());
  NT2_TEST_EQUAL(lo(boost::simd::One<T>()), boost::simd::Zero<dtype>());
  NT2_TEST_EQUAL(lo(boost::simd::Zero<T>()), boost::simd::Zero<dtype>());
} // end of test for floating_

NT2_TEST_CASE_TPL ( lo_int64__1_0,  BOOST_SIMD_INT_64_TYPES)
{

  using boost::simd::lo;
  using boost::simd::tag::lo_;
  typedef typename boost::dispatch::meta::as_integer<T,unsigned>::type ir_t;
  typedef typename boost::dispatch::meta::downgrade<ir_t>::type dtype;
  typedef typename boost::simd::meta::scalar_of<ir_t>::type scal;
  typedef typename boost::dispatch::meta::as_integer<T>::type iT;
  typedef typename boost::dispatch::meta::call<lo_(T)>::type r_t;
  typedef typename boost::simd::meta::scalar_of<r_t>::type sr_t;
  typedef typename boost::simd::meta::scalar_of<r_t>::type ssr_t;
  typedef dtype wished_r_t;


  // return type conformity test
  NT2_TEST( (boost::is_same < r_t, wished_r_t >::value) );
  std::cout << std::endl;

  // specific values tests
  NT2_TEST_EQUAL(lo(boost::simd::One<T>()), boost::simd::One<dtype>());
  NT2_TEST_EQUAL(lo(boost::simd::Zero<T>()), boost::simd::Zero<dtype>());
} // end of test for int64_

NT2_TEST_CASE_TPL ( lo_int32__1_0,  BOOST_SIMD_INT_32_TYPES)
{

  using boost::simd::lo;
  using boost::simd::tag::lo_;
  typedef typename boost::dispatch::meta::as_integer<T,unsigned>::type ir_t;
  typedef typename boost::dispatch::meta::downgrade<ir_t>::type dtype;
  typedef typename boost::simd::meta::scalar_of<ir_t>::type scal;
  typedef typename boost::dispatch::meta::as_integer<T>::type iT;
  typedef typename boost::dispatch::meta::call<lo_(T)>::type r_t;
  typedef typename boost::simd::meta::scalar_of<r_t>::type sr_t;
  typedef typename boost::simd::meta::scalar_of<r_t>::type ssr_t;
  typedef dtype wished_r_t;


  // return type conformity test
  NT2_TEST( (boost::is_same < r_t, wished_r_t >::value) );
  std::cout << std::endl;
  // specific values tests
  NT2_TEST_EQUAL(lo(boost::simd::One<T>()), boost::simd::One<dtype>());
  NT2_TEST_EQUAL(lo(boost::simd::Zero<T>()), boost::simd::Zero<dtype>());
} // end of test for int32_

NT2_TEST_CASE_TPL ( lo_int16__1_0,  BOOST_SIMD_INT_16_TYPES)
{

  using boost::simd::lo;
  using boost::simd::tag::lo_;
  typedef typename boost::dispatch::meta::as_integer<T,unsigned>::type ir_t;
  typedef typename boost::dispatch::meta::downgrade<ir_t>::type dtype;
  typedef typename boost::simd::meta::scalar_of<ir_t>::type scal;
  typedef typename boost::dispatch::meta::as_integer<T>::type iT;
  typedef typename boost::dispatch::meta::call<lo_(T)>::type r_t;
  typedef typename boost::simd::meta::scalar_of<r_t>::type sr_t;
  typedef typename boost::simd::meta::scalar_of<r_t>::type ssr_t;
  typedef dtype wished_r_t;


  // return type conformity test
  NT2_TEST( (boost::is_same < r_t, wished_r_t >::value) );
  std::cout << std::endl;

  // specific values tests
  NT2_TEST_EQUAL(lo(boost::simd::One<T>()), boost::simd::One<dtype>());
  NT2_TEST_EQUAL(lo(boost::simd::Zero<T>()), boost::simd::Zero<dtype>());
} // end of test for int16_
